### ISSUE TITLE
RetinaNet code example 

### ISSUE THREAD

**Issue by Suzan009** [keras-team/keras#15769]
@srihari-humbarwadi 
I'm following the keras code example on object detection, [retina-net](https://keras.io/examples/vision/retinanet/), and a few things are not clear to me. 

1. anchor box: It uses an anchor box by assuming the target shape **(xmin, ymin, width, height)**. Now, what if the target is **(xmin, ymin, xmax, ymax)**? [Code](https://keras.io/examples/vision/retinanet/#Implementing%20Anchor%20generator). 

```python
class AnchorBox:
    """Generates anchor boxes.
    This class has operations to generate anchor boxes for feature maps at
    strides `[8, 16, 32, 64, 128]`. Where each anchor each box is of the
    format `[x, y, width, height]`.
....
````

What do I need to change of the above code? Direct [Code](https://keras.io/examples/vision/retinanet/#Implementing%20Anchor%20generator) link. 

2. In data pipelines, what is the use of padding and the **values**, like below: 

```python
train_dataset = train_dataset.padded_batch(
    batch_size=batch_size, 
   padding_values=(0.0, 1e-8, -1), 
   drop_remainder=True
)
```

and why we use 

```python
train_dataset.apply(tf.data.experimental.ignore_errors())

```

**Comment by Suzan009** [comment]
@srihari-humbarwadi great work btw.

**Comment by sachinprasadhs** [comment]
It is basically used when creating the Dataset from another Dataset, in that process if there is any element which caused error it will be dropped silently.
You can see more details with example [here](https://www.tensorflow.org/api_docs/python/tf/data/experimental/ignore_errors). Thanks!

**Comment by Suzan009** [comment]
Understood. It's safe during training but one drawback I think is that we may never know if there are any corrupt elements in my data set as there will not raise any warning. 

Could you address the other issue that I mentioned above?

**Comment by old-school-kid** [comment]
Hi @Suzan009 
Regarding your first issue, you can design and train you own model to predict (xmin, ymin, xmax, ymax) rather than (xmin, ymin, width, height). But the model retina-net was designed to predict (xmin, ymin, width, height). So you can convert your (xmin, ymin, xmax, ymax) labels to (xmin, ymin, width, height) by using the function `convert_to_xywh` and revert back to (xmin, ymin, xmax, ymax) using `convert_to_corners` function. Both the function definitions are given in the same piece of code itself. 
Hope this helps!

**Comment by srihari-humbarwadi** [comment]
I hope that the comments above do resolve your queries @Suzan009, and for 
>In data pipelines, what is the use of padding and the values, like below:

The model is trained on variable image sizes. Hence, to accommodate images of different sizes in a batch, we pad the images and labels.
 - images are padded with `0s`
 - boxes are padded with a small float value `1e-8`
 - and class ids are padded with `-1` (ignore label)

**Comment by Suzan009** [comment]
@old-school-kid thanks, It helps. 
@srihari-humbarwadi thanks for the clarification. I think you should add this information to your code example. However, I noticed that you are still working on this [repo](https://github.com/srihari-humbarwadi/retinanet-tensorflow2.x), and that's why I think it would be great to add more educative section in your code example. Great work, 👍  

Do you mind giving some suggestions regarding anchor optimization for custom data set? Is it possible to auto-tune it by itself based on the custom data?

**Comment by Suzan009** [comment]
And is it possible to visualize the generated anchor boxes conveniently?

**Comment by old-school-kid** [comment]
> Do you mind giving some suggestions regarding anchor optimization for custom data set? Is it possible to auto-tune it by itself based on the custom data? (pointer)
As fas as I know [this](https://arxiv.org/abs/1807.00980) is SOTA mechanism when it comes to anchor boxes. And as always K-Means also produces quite satisfatory results. You can also try out IoU loss.
